### PR TITLE
Bug 1914053: whereabouts: add ip-reconciler cronjob

### PR DESCRIPTION
**Pull request by maiqueb**
Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

### CODE REVIEW

File: bindata/network/multus/multus.yaml
Comment (maiqueb) [reply bindata/network/multus/multus.yaml:593]: Ah! I found it available on the correct path, which is actually referenced by the `SystemCNIConfDir` variable:
https://github.com/openshift/cluster-network-operator/blob/87d4207091c6a6aed36ca9cf6f190c79c9a8d456/pkg/network/multus.go#L14

But, the file is available on host with r/w permissions only to the root user:
```
sh-4.4# ls -la /etc/kubernetes/cni/net.d/whereabouts.d/
total 16
drwxr-xr-x. 2 root root    60 Oct 14 09:14 .
drwxr-xr-x. 4 root root    65 Oct 14 09:30 ..
-rw-r--r--. 1 root root   162 Oct 14 09:29 whereabouts.conf
-rw-------. 1 root root 11559 Oct 14 09:29 whereabouts.kubeconfig

```
Which causes our humble pod to not be able to access it:
```
oc logs -nopenshift-multus ip-reconciler-27236755--1-zzwlw
2021-10-14T09:55:44Z [debug] NewReconcileLooper - Kubernetes config file located at: /host/etc/cni/net.d/whereabouts.d/whereabouts.kubeconfig
2021-10-14T09:55:44Z [error] failed to instantiate the Kubernetes client: Error loading config file "/host/etc/cni/net.d/whereabouts.d/whereabouts.kubeconfig": open /host/etc/cni/net.d/whereabouts.d/whereabouts.kubeconfig: permission denied
2021-10-14T09:55:44Z [error] failed to create the reconcile looper: failed to instantiate the Kubernetes client: Error loading config file "/host/etc/cni/net.d/whereabouts.d/whereabouts.kubeconfig": open /host/etc/cni/net.d/whereabouts.d/whereabouts.kubeconfig: permission denied

```

File: bindata/network/multus/multus.yaml
Comment (maiqueb) [reply bindata/network/multus/multus.yaml]: You mean I do not need to mount the entire cni config folder, right ? Just the particular k8s config file. 

That's a good point. Once we decide how we want to do this, I'll update it to point at this particular file.

Comment (dougbtv) [reply bindata/network/multus/multus.yaml]: +1, you can use just the file.

File: bindata/network/multus/multus.yaml
Comment (dougbtv) [reply bindata/network/multus/multus.yaml]: Hrmmm, I'm a little torn. So, all the other CNI components herein are privileged. So it's not... weird. 

Maybe it's even appropriate. Because if we make that file more easily readable, then, that service account credentials are exposed.

**Comment by dougbtv** [comment]
/title Bug 1914053: whereabouts: add ip-reconciler cronjob

**Comment by dougbtv** [comment]
/retitle Bug 1914053: whereabouts: add ip-reconciler cronjob

File: bindata/network/multus/multus.yaml
Comment (danwinship) [inline bindata/network/multus/multus.yaml:468]: Can you add a comment before this explaining what it means? I always forget which cronjob field is which...

File: bindata/network/multus/multus.yaml
Comment (danwinship) [inline bindata/network/multus/multus.yaml:473]: This seems maybe incorrect for this job? I forget exactly what this field does, but it seems like a periodic cleanup job is not "system critical".

File: bindata/network/multus/multus.yaml
Comment (danwinship) [inline bindata/network/multus/multus.yaml:475]: `privileged` is only defined in the container-level `securityContext`, not the pod-level. So this has no effect

File: bindata/network/multus/multus.yaml
Comment (maiqueb) [reply bindata/network/multus/multus.yaml:491]: The reasons for it are described in https://github.com/openshift/cluster-network-operator/pull/1207/commits/fdf46947684e41f7b3ae97dcda377d2924a8ca8c#r728837192 . 

This pod does **not** require to be privileged - it could regenerate the kubeconfig file instead. 

Your call.

File: bindata/network/multus/multus.yaml
Comment (danwinship) [reply bindata/network/multus/multus.yaml:474]: As I said, I don't remember where the exact definition of how to deal with this field is... (maybe something in openshift/enhancements/dev-guides says? If not, probably there _should_ be something there).

But, for example, if a node is at max capacity, and it's time to run this job, do you want kubernetes to kill user workload pods to make room for this pod? I'm guessing no, in which case this probably needs to have different priority characteristics than most other network-related pods.

**Comment by danwinship** [comment]
You don't need an actual kubeconfig when you're running as a pod; call `rest.InClusterConfig()` from `k8s.io/client-go/rest` and it will use `KUBERNETES_SERVICE_HOST` and the service token, etc, to generate an appropriate kubeconfig for you. (You can just make the existing `kubernetes.NewClient` in whereabouts try both ways; `InClusterConfig` will return a specific error you can check for if you call it from outside a pod.)

**Comment by dougbtv** [comment]
/hold

waiting for https://github.com/openshift/whereabouts-cni/pull/73 to be merged so we have the downstream integration

File: bindata/network/multus/multus.yaml
Comment (danwinship) [inline bindata/network/multus/multus.yaml:469]: Oh, huh, I was thinking it meant "run 5 minutes after the hour", not "run every 5 minutes".

In the long term we should think about whether it makes more sense to just have some existing long-running pod handle this, rather than spinning up a pod every 5 minutes... (Maybe CNO could do it?)

File: bindata/network/multus/multus.yaml
Comment (danwinship) [inline bindata/network/multus/multus.yaml:474]: Did you actually find any documentation about this field and how to pick a value for it? (We should probably link to that from the wiki so that next time we add something we know where to look...)

**Review by danwinship — APPROVED** [review]
/lgtm
though I'd still like more info on `priorityClassName`...

File: bindata/network/multus/multus.yaml
Comment (maiqueb) [reply bindata/network/multus/multus.yaml:469]: Turning whereabouts into a thick plugin (long running pod) and migrate the ip cleanup reconcile loop there is one of our goals. I can't speak for **when** though... But it's on our radar.

File: bindata/network/multus/multus.yaml
Comment (maiqueb) [reply bindata/network/multus/multus.yaml:474]: Unfortunately I didn't. 

I went with the intermediate value. Having said that, I agree that it would be very helpful for the next time it is needed.

**Pull request merged**
